### ISSUE TITLE
[FEATURE PROPOSAL] Berry Farm Overhaul 

### ISSUE THREAD

**Issue by HLXII** [pokeclicker/pokeclicker#902]
# Abstract
The current berry system is fairly removed from the rest of the game. The only connect it has to other mechanics is getting Farm points to then spend on Master balls, or completing quests for QP. This process is also quite tedious, since the Farm mechanic requires constant attention since there is no passive gain implemented in the mechanic.  What I mean by this is that you can passively gain Gold and DP by sitting in a Route with Poke balls. However, to get Farm points, you have to be active every 1-6 minutes.

My goal for this proposal is to integrate the Farm mechanic into other systems, so that the player has more incentive to use the mechanic, and to make the experience a bit richer for the player, rather than clicking Plant All, waiting for a notification, then Harvest All and repeat.

#  Berry Farm

## [Berry Types](https://bulbapedia.bulbagarden.net/wiki/Berry#Generation_III_onwards)

The number of berry types will expand to all canon berries (possibly excluding Gen VI ones, as those don't have [artwork](https://bulbapedia.bulbagarden.net/wiki/Berry#Artwork)). We'll keep the same BerryDistribution constant and continue to drop the same 8 berries implemented. The rest of the berry types will be gained by either:

1. Mutations (see Mutations section)
2. Achievements (?)
   *  I was thinking one berry type would be gained by getting all other berry types, but not so sure about this one.

### Berry Generations

There are a total of 67 berries in canon pokemon. I've separated them into five different generations.

First Gen | Second Gen | Third Gen | Fourth Gen (Typed Berries) | Fifth Gen
------------ | ------------- | ------------- | ------------- | -------------
Cheri |  Persim | Pomeg | Occa | Liechi
Chesto | Razz | Kelpsy | Passho | Ganlon
Pecha | Bluk | Qualot | Wacan | Salac
Rawst | Nanab | Hondew | Rindo | Petaya
Aspear | Wepear | Grepa | Yache | Apicot
Leppa | Pinap | Tamato | Chople | Lansat
Oran | Figy | Cornn | Kebia | Starf
Sitrus | Wiki | Magost | Shuca | Micle
_ | Mago | Rabuta | Coba | Custap
_ | Aquav | Nomel | Payapa | Jaboca
_ | Iapapa | Spelon | Tanga | Rowap
_ | Lum | Pamtre | Charti | Kee
_ | _ | Watmel | Kasib | Maranga
_ | _ | Durin | Haban | Enigma
_ | _ | Belue | Culbur
_ | _ | _ | Babiri
_ | _ | _ | Chilan
_ | _ | _ | Roseli

The generations go in order of how the player will generally unlock them. The second generation will have mutation requirements that require the first gen. The third gen a combination of first and second, etc. The generations will also be an indicator for how long the berries take to grow. The early generations will be quick to grow, and as the player unlocks more, the berries will take longer. I plan for the Enigma berry to take a couple of days. This fits with the playstyle of the player, as the farther they get, the more idling they will do, so it fits to have a farm idle with you.

## Unlocking Plots

Using Farm points is a bit uninteresting, since most people (if they're trying to go fast), will look at the points/sec ratio and choose Cheri berries, or if they're lazy, choose Sitrus. I propose changing the unlock method to use the berries themselves. That way the player has a reason to farm other berries. This will go down the list of the first 8 berries (10 Cheri, 15 Chesto,.. etc). After the first 8 are unlocked, the player will have to start using Mutations to get the new berry types. This promotes growing all berry types (at least until the entire plot is unlocked).

## Growth Improvements

The current berry growth is determined by two things: The Sprayduck, and the inherent growth time. This is <i>fine</i>, though it could be a bit more complex.

I propose splitting the growth times into more properties. So the time from Seed > Bud > Sprout > Bloom > Berry will have varying times depending on the Berry type. This might be a little too complex, so I'm not as married to this. However, the stages will be important for the Mutations section.

Berries will also have a death timer after being ripe, also a property of the berry type. Some berries can last quite a while in the Berry stage, while some die quickly.

Berries will also have a chance of dropping berries on death. This gives at least some passive generation, where the player doesn't need to baby sit the Farm waiting on notifications. The dropped berries also have a chance of replanting themselves, also giving some passive generation.

These probabilities will also be affected by new Oak items specific for berries. (See Oak Item section).

We'll also remove gaining Money through farming. This doesn't really make sense (why are you gaining money by harvesting berries?), so you'll only gain Farm points.

## Additional Properties

To add additional complexity (not for growing but for use cases), we'll add some additional properties to the berry types. These will be used later on with NPCs.

### [Flavor](https://bulbapedia.bulbagarden.net/wiki/Flavor#List_of_Berries_by_flavor)

Each berry will have the flavor assigned to it in the game. This can be expanded quite a bit depending on how expansive the berry system will be. Largest expansion would be creating a kitchen type location where you can combine berries into consumables that give some kind of bonus.

The flavor property can also have a hand in explaining some mutations. Maybe a mutation requirement might be three spicy berries near each other.

### [Color](https://bulbapedia.bulbagarden.net/wiki/Berry_color#List_of_Berries_by_color)

The color of the berry is also a property assigned in the canon games. Like flavor, this can be used as an explanation for some mutation requirements, however it could also be used as a category for handling the daily trades for the NPCs. For instance, maybe one trade will always require a Blue, Red, and Green berry (with their quantities varying). 

## [Mulch](https://bulbapedia.bulbagarden.net/wiki/Mulch)

Mulch will be able to be bought from the Berry Master in Goldenrod City. In canon, this was sold in the [flower shop](https://bulbapedia.bulbagarden.net/wiki/Goldenrod_City#Flower_shop), but It'll be easier to group them together (See NPC section for more details).

There will be 4 types of Mulch (and possibly 8 including the Kalos ones. Those might be locked until Kalos, or they might be used for my Apricorn idea, this is still undecided). I've decided on using the Kalos ones for Berry Farms, and the HGSS ones for Apricorns.

Mulch | Description
------------ | -------------
Boost Mulch |  Increases growth rate
Rich Mulch | Increases Berry production
Surprise | Increases Mutation rate (see Mutation section)
Amaze | Has all three effects, however a weaker version of them

Only one mulch may be applied at a time. One decision to be made is if the mulches should be applied per farm plot, or as an effect for the entire farm.

## [Mutations](https://bulbapedia.bulbagarden.net/wiki/Berry_fields_(Kalos)#Mutation)

Berries will have the possibility of mutating into another type, depending on some factors. Some examples of mutations can be seen in the canon wiki, however they'll be slightly altered for Pokeclicker.

Mutations can depend on two factors:

 * Growth Stage
    * Example: An Oran berry in the Bloom stage next to a Pecha berry in Berry stage has a chance of producing a Qualot berry.
 * Positioning
    * Example: An Aspear berry must be diagonal to a Leppa berry to create a Hondew berry.

Players will have to plan out their timings and the positioning of Berry plants in order to create new Berry types. There is also a Berry Master in Kanto that will give random hints for possible Mutation requirements (if the players are too lazy to look at the wiki).

## Safari Fuel

There was previously a suggestion to use the berries in the Safari for bait instead of just having a bait option. We can implement this suggestion here as well. Specifically, the Razz berry and Nanab berry have canon uses for making pokemon easier to capture and making the move less respectively. This only applies to Pokemon Let's Go, but it can be morphed into Pokeclicker as well. Currently in my Mutations theory-crafting, I have these two berries in the first generation of mutations. So once the player has access to the farm, these berries should be among the first that are obtained. There will need to be hints in both the Safari Modal (before actually spending QP), as well as the Kanto Berry Master on the uses of these berry types.

## Attracting Pokemon

Each Berry type will also have their own pool of Pokemon that are attracted to that type of berry. This could be another avenue of catching pokemon, as it's another passive way of gaining additional dungeon points. In addition, certain pokemon can be exclusively gained from the wandering pokemon mechanic. The main example of this would be Combee, which currently is being sold in a shop for QP. Moving Combee's location to the farm would add another reason for the player to invest time into the farm mechanic.

# NPCs

## [Berry Masters](https://bulbapedia.bulbagarden.net/wiki/Berry_Master)

Each region will have their own Berry Master. These will be NPCs in towns that the player can interact with to open a Shop modal.

### Kanto

There is a [Berry Powder man](https://bulbapedia.bulbagarden.net/wiki/Cerulean_City#Berry_Powder_man) in Cerulean City. I'm morphing that into the first Berry Master in Pokeclicker. In canon, he sells supplements, but since Protein hasn't been implemented yet, I've decided to ignore that (later in development, this NPC could also have trades for supplements). This Berry Master, when talked to, will give hints to possible mutations for the berries, changing the hint every day. We can also tailor this to only show mutations that the player hasn't discovered yet.

### Johto

The Berry Master can be found in Goldenrod City. This Berry Master will sell Mulch, as well as have additional trades. These trades will consist of more variations in berry types (as by this point, the player should have noticed some mutations occuring, or read the wiki). The trades for this Berry Master will consist of Battle Items, and sometimes Evolution items, and will also be rotated every day. The Evolution items of course will be limited based on the current max region of the player. One possible implementation could also skew the probabilities of the trades to fit with what the player currently requires. For example, if the player is currently in Hoenn, the Berry Master might have a higher chance of trading Deep Sea Scales or such compared with Evolution items introduced in Hoenn or Johto. They will also sell the Squirtbottle Oak Item.

The modal for the Berry Master will consist of two tabs: The Shop tab, and the Trade tab. The shop tab will act like regular shop tabs and contain Mulches for Farm Points, and the SquirtBottle for Quest Points. The trade tab will consist of 4 trades. The first three trades will be for Battle Items, and they'll require 2 random berry types, one from the first generation, and one from the second. The amount for each will also vary. The last trade will be for three berry types, from the first, second, and third gen berries. This will trade for a random evolution item.

### Hoenn 

The Berry Master here lives in Route 123. If we don't want to code in a location just for the Berry Master, we can instead move him to Mauville or Fortree. Anyways, this Berry Master has a shop modal that trades rare berries for Underground items. Rare berries being those only unlocked by mutations. They will also sell berries for Farm points, however only berries that the player has already obtained (so they'll still need to use Mutations). They will also sell the Sprinklotad Oak Item.

The modal will be the same as the Johto Berry Master (we could possibly also sell Mulches here so the player doesn't always have to go to Johto). This Berry Master have the Sprinklotad Oak Item instead. On the trade tab, there will be three separate trades (much like the Underground). The trades will consist of only third and fourth gen berries. 

### Sinnoh

The Berry Master here lives in Route 208, but we can morph that into Hearthome City. The shop modal here is one trade every day (in that you can only trade once then the shop is out of stock). Here, a Master ball can be traded for, however the trades will be quite expensive. However, since the trades are generated using a seed (much like the Daily trades in the Underground), players can plan out their berry growing.

This Berry Master will only have a trade tab in the modal (aka just a trade modal). The trade will always be for a Masterball with a stock of 1 (reset either weekly or daily). It will require 5 types of berries, all chosen from the five generations (amounts scaling of course, so a lot more first gen berries and maybe 5-10 fifth gen).

### Beyond

There aren't any other Berry Masters in the game (except for Kalos, but I'm ignoring that for now), so those are the only Berry Masters. Also this expands the mechanic to 4 separate regions, which I think is enough. This mechanic also gives the player incentive to use the Town Map, as there won't be shortcuts for the Berry Masters.

## [Pokeblocks](https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9block)

Pokeblocks are created from berries in Hoenn to be used in Contests. Since there isn't an idea for a Contest mechanic yet, I'm reusing them as another upgrade item type.

The Pokeblock case key item can be found in canon in the Contest Hall in Slateport. It will be able to be purchased from the the Pokeblock NPC found in Slateport. After purchasing, talking to the Pokeblock NPC will bring up the Pokeblock modal. Here the player can use their berries in the Berry Blender to create Pokeblocks. There will be 18 different pokeblock types to match every Pokemon type. They will be created by using the fourth gen berries (which match to all 18 types), as well as some filler berries (probably from the lower types. not sure of the actual recipes yet). These could act like additional Battle Items that increase the attack of that specific Pokemon type.

We could also add in Poffins later on, but I don't want just a duplicate of Pokeblocks.

# Oak Items

There is currently only one Oak Item for berries, that being the Sprayduck. Since this mechanic is so broad now, I'm just going to add 2 more 😄 .

1. Squirtbottle - Bought from the Johto Berry Master with Farm points. Increases Mutation chance. Level up by having mutations occur.
2. Sprinklotad - Bought from the Hoenn Berry Master with Farm points. Increases replant chances. Level up by having replants occur.

# Balance

This proposal doesn't have any numbers yet, so I thought I'd add a couple of ballpark values.

## Growth

The currently implemented berries are fully grown in 1-6 minutes. This is fairly uninteresting, and also requires constant attention. I was thinking 5-10 minute growth rates for the first 8 berries, so that the player can focus their attention on the new mechanic that they've just unlocked. For later berries gained from mutations, the growth rate will be slower, taking 10-60 minutes, as the player will be more idle anyways for grinding. The end game mutation berries will take a couple of hours to grow.

## Mutation

The first round of mutations should be fairly easy to obtain, so that near end of Kanto, most of the first round should've been unlocked. Later mutations should be harder to get, since the player has more space in the plot, and also because they'll be idling more.

## Amounts

This will be quite hard to balance, since most players have thousands of Cheri berries already. Maybe that will just be a wash, and we won't sell anything important using first gen berries. The later rounds of mutation will be more expensive, since they'll be used to buy Evolution Items and Master balls. Thankfully this is okay with current saves since they'll be new berries that will have to be farmed.

# Conclusion

This is a lot of information/ideas to cover, and I doubt most or if any of it will be implemented. However, I believe that the Farm mechanic currently as it is now doesn't have much going for it. It is only a bit better than the Safari Zone (which is a nightmare). Hopefully some of the ideas mentioned here are considered for improving this mechanic so that the player has something to fiddle with while their Pokemon trying to catch Mew or Suicune.

For what it's worth, I've also started a branch for a mockup of these changes: https://github.com/HLXII/pokeclicker/tree/berry-overhaul

# Coding Overview

## High Priority
*  [x] Berries
   * [x] - Implement Second Generation
   * [x] - Implement Third Generation
   * [x] - Implement Fourth Generation
   * [x] - Implement Fifth Generation
*  [ ] Mutations
   * [x] - Implement Second Generation Mutations
   * [x] - Implement Third Generation Mutations
   * [x] - Implement Fourth Generation Mutations
   * [ ] - Implement Fifth Generation Mutations
   * [x] - Implement Side Mutations
*  [x] NPCs
   * [x] - Kanto Berry Master
      * [x] - Setup mutation messages
      * [x] - Setup random generation
      * [x] - Setup NPC
   * [x] - Johto Berry Master
      * [x] - Setup Berry Master Modal
         * [x] - Refactor Shop to handle Berry Masters
         * [x] - Add Berry Trade tab
         * [x] - Setup Oak Item purchase
      * [x] - Set up random deal generation
   * [x] - Hoenn Berry Master
      * [x] - Set up random deal generation
   * [x] - Sinnoh Berry Master
      * [x] - Set up random deal generation
## Medium Priority
- [x] - Grab new images for berries
- [x] - Add Flavor UI in BerryDex
- [x] - Update Safari with new berries
   * [x] - Update Modal
   * [x] - Setup berry properties
   * [x] - Update backend to handle new berries affecting chances
- [x] - Wandering Pokemon
   * [x] - Setup berry properties
   * [x] - Setup random generation
   * [x] - Setup notification
   * [x] - Add pokemon to party
- [x] - Add Berry Aura effects
- [x] - Create new images for Squirtbottle and Sprinklotad (or fake them with Dream World icons)
- [x] - Add berry description (including additional details on new behavior)
- [x] - Get sprites for Gen VI berries [here ](https://www.rebornevo.com/forums/topic/41667-what-do-the-kee-berry-and-maranga-berry-trees-look-like/)is someone's art we can steal. Need to convert into the animated PNG
- [x] - Add Shovel item to remove Berry plants
 - [x] - Update Berry Quests
## Low Priority
- [x] - Update Sprayduck to gain experience based on berry type harvested
- [ ] - Possibly store Kanto Berry Master hints in BerryDex
- [ ] - Implement PokeBlocks (not sure of use case yet)
- [ ] - Implement loadout system
- [x] - Update Berry tooltip to format info better. Maybe include aura emitted and nearby berry auras affecting plot.
- [ ] - Add mulch multipliers button somewhere (for 10x application, etc)

**Comment by Aegyo** [comment]
This all looks very interesting, the farm has been needing an overhaul for a while.
I have an addition, to encourage a variety of berries planted - Each berry has an effect on its neighbor, unless the neighbor is the same kind of berry. These effects could be positive (faster growth time, higher yield, increased mutation chance) or negative on some of the more lucrative berries (whatever lucrative means in this new overhaul). Hopefully that would encourage lots of experimentation trying to create the best layout, and the best layout would probably be different depending on your goal (lots of farm points, farming a specific berry for the safari, increasing your mutation chances)

**Comment by KrushGames** [comment]
Jessica that sounds cool, but only if we can save the loadout we deem or have an option for "replant loadout" (or something similar)

**Comment by Aegyo** [comment]
If it seems popular, I'm sure we could have layouts saved and even stored/loaded by a short string you can share with people.

**Comment by Aegyo** [comment]
In the same kind of area, #659 suggests having the oak items gain more exp with the better berries

**Comment by HLXII** [comment]
Merged
